In [1]:
import pandas as pd 
df = pd.read_csv(r"C:\Users\Rasulbek_Ruzmetov\Desktop\SML_R_Project\Data\Preprocessing_Data\Data_Preprocessing.csv")

In [2]:
df

work_year  job_title     salary  salary_currency  salary_in_usd  \
0       0.533424  -0.719710   0.018352         0.220287       0.140113   
1       0.533424  -0.719710  -0.159937         0.220287      -0.418953   
2       0.533424   0.575812  -0.327013         0.220287      -0.942856   
3       0.533424   0.575812  -0.478321         0.220287      -1.417317   
4       0.533424   0.661044   0.152644         0.220287       0.561217   
...          ...        ...        ...              ...            ...   
40327  -6.193918  -0.020810   1.074657         0.220287       3.452396   
40328  -4.512083   1.274712  -0.069094         0.220287      -0.134095   
40329  -6.193918  -0.020810  -0.270675         0.220287      -0.766197   
40330  -6.193918  -1.265192  -0.292586         0.220287      -0.834904   
40331  -4.512083  -0.020810  29.944529        -3.133669      -0.908214   

       employee_residence  remote_ratio  company_location  \
0                0.304715     -0.542407          0.303171   
1                0.304715     -0.542407          0.303171   
2                0.304715      1.860795          0.303171   
3                0.304715      1.860795          0.303171   
4                0.304715      1.860795          0.303171   
...                   ...           ...               ...   
40327            0.304715      1.860795          0.303171   
40328            0.304715      1.860795          0.303171   
40329            0.304715      1.860795          0.303171   
40330            0.304715      1.860795          0.303171   
40331           -2.132674      0.659194         -2.080539   

       experience_level_EN  experience_level_EX  experience_level_MI  \
0                -0.307153            -0.150043            -0.640438   
1                -0.307153            -0.150043            -0.640438   
2                -0.307153            -0.150043             1.561432   
3                -0.307153            -0.150043             1.561432   
4                -0.307153            -0.150043            -0.640438   
...                    ...                  ...                  ...   
40327            -0.307153            -0.150043            -0.640438   
40328            -0.307153            -0.150043             1.561432   
40329             3.255711            -0.150043            -0.640438   
40330             3.255711            -0.150043            -0.640438   
40331            -0.307153            -0.150043            -0.640438   

       experience_level_SE  employment_type_CT  employment_type_FL  \
0                 0.814945           -0.036953           -0.018634   
1                 0.814945           -0.036953           -0.018634   
2                -1.227076           -0.036953           -0.018634   
3                -1.227076           -0.036953           -0.018634   
4                 0.814945           -0.036953           -0.018634   
...                    ...                 ...                 ...   
40327             0.814945           -0.036953           -0.018634   
40328            -1.227076           -0.036953           -0.018634   
40329            -1.227076           -0.036953           -0.018634   
40330            -1.227076           27.061210           -0.018634   
40331             0.814945           -0.036953           -0.018634   

       employment_type_FT  employment_type_PT  company_size_L  company_size_M  \
0                0.056646           -0.038599       -0.197964        0.211123   
1                0.056646           -0.038599       -0.197964        0.211123   
2                0.056646           -0.038599       -0.197964        0.211123   
3                0.056646           -0.038599       -0.197964        0.211123   
4                0.056646           -0.038599       -0.197964        0.211123   
...                   ...                 ...             ...             ...   
40327            0.056646           -0.038599        5.051412       -4.736582   
40328            0.056646      

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40332 entries, 0 to 40331
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   work_year            40332 non-null  float64
 1   job_title            40332 non-null  float64
 2   salary               40332 non-null  float64
 3   salary_currency      40332 non-null  float64
 4   salary_in_usd        40332 non-null  float64
 5   employee_residence   40332 non-null  float64
 6   remote_ratio         40332 non-null  float64
 7   company_location     40332 non-null  float64
 8   experience_level_EN  40332 non-null  float64
 9   experience_level_EX  40332 non-null  float64
 10  experience_level_MI  40332 non-null  float64
 11  experience_level_SE  40332 non-null  float64
 12  employment_type_CT   40332 non-null  float64
 13  employment_type_FL   40332 non-null  float64
 14  employment_type_FT   40332 non-null  float64
 15  employment_type_PT   40332 non-null 

In [20]:
import plotly.express as px
import pandas as pd

# Salary bilan correlation hisoblash
corr_matrix = df.corr()

# Salary ustunini olish
salary_corr = corr_matrix[["salary"]].reset_index()
salary_corr.columns = ["Feature", "Correlation"]

# Heatmap yaratish
fig = px.bar(salary_corr.sort_values(by="Correlation", ascending=False),
             x="Feature", y="Correlation",
             title="Salary bilan boshqa ustunlar o‘rtasidagi correlation",
             text="Correlation")

fig.update_traces(texttemplate="%{text:.2f}", textposition="outside")
fig.update_layout(yaxis=dict(range=[-1, 1]))
fig.show()


In [5]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR

from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

# ===============================
# 1. Ma'lumotlarni tayyorlash
# ===============================
X = df.drop(columns=["salary"])
y = df["salary"]

# Categorical ustunlar bo'lsa, ularni one-hot encoding qilish
X = pd.get_dummies(X, drop_first=True)

# Train/Test bo‘lishi
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Ma'lumotlarni scaling qilish
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# ===============================
# 2. Regression modellar ro‘yxati
# ===============================
models = {
    "Linear Regression": LinearRegression(),
    "Decision Tree": DecisionTreeRegressor(random_state=42),
    "Random Forest": RandomForestRegressor(random_state=42),
    "Gradient Boosting": GradientBoostingRegressor(random_state=42),
    "Support Vector Regressor": SVR()
}

# ===============================
# 3. Model natijalarini saqlash uchun ro‘yxat
# ===============================
results_list = []

# ===============================
# 4. Modelni o‘rgatish va baholash
# ===============================
for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    r2 = r2_score(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)
    mae = mean_absolute_error(y_test, y_pred)

    results_list.append({"Model": name, "R2": r2, "MSE": mse, "MAE": mae})

# ===============================
# 5. Natijalarni DataFrame ga aylantirish
# ===============================
results = pd.DataFrame(results_list)
print(results.sort_values(by="R2", ascending=False))


                      Model        R2       MSE       MAE
1             Decision Tree  0.708436  0.171542  0.013969
2             Random Forest  0.667047  0.195894  0.023860
3         Gradient Boosting  0.611383  0.228644  0.032204
4  Support Vector Regressor  0.195257  0.473472  0.062556
0         Linear Regression  0.157048  0.495953  0.101052


# 📊 Asosiy Xulosalar — Regression Modellarini Taqqoslash

## **1. Eng yaxshi model**
- **Decision Tree Regressor** eng yaxshi natija berdi — **R² = 0.708**, bu model ma’lumotlarning ~70% variansiyasini tushuntira oladi.
- **MSE** va **MAE** eng past bo‘lib, modelning taxminlari juda aniq ekanini ko‘rsatadi.

---

## **2. Ikkinchi o‘rinda**
- **Random Forest Regressor** (**R² = 0.667**) yaxshi ishladi, lekin Decision Tree’dan biroz past.
- Bu model umumiy stabillik va yuqori aniqlik uchun yaxshi variant hisoblanadi.

---

## **3. Gradient Boosting Regressor**
- **R² = 0.611** bilan o‘rtacha natija berdi.
- **MSE** va **MAE** qiymatlari biroz yuqori, shuning uchun murakkabligi inobatga olinib, bu model aniqroq sozlash (**hyperparameter tuning**) bilan yaxshilanishi mumkin.

---

## **4. Kuchsiz ishlagan modellari**
- **Support Vector Regressor** (**R² = 0.195**) va **Linear Regression** (**R² = 0.157**) natijalari juda past.
- Bu, ehtimol, ma’lumotlarda murakkab no-chiziqli bog‘lanishlar borligini va oddiy chiziqli yoki standart SVR model ular uchun yetarli bo‘lmasligini ko‘rsatadi.


In [ ]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

# ===============================
# Ma'lumot tayyorlash
# ===============================
X = df.drop(columns=["salary"])
y = df["salary"]

X = pd.get_dummies(X, drop_first=True)

# Train/Test bo‘lishi
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# ===============================
# Model tayyorlash
# ===============================
model = DecisionTreeRegressor(random_state=42)
model.fit(X_train, y_train)

# ===============================
# Taxmin qilish va baholash
# ===============================
y_pred = model.predict(X_test)

r2 = r2_score(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)

print("📌 Decision Tree Regressor — Yakuniy Natijalar:")
print(f"R2: {r2:.4f}")
print(f"MSE: {mse:.4f}")
print(f"MAE: {mae:.4f}")

📌 Decision Tree Regressor — Yakuniy Natijalar:
R2: 0.7084
MSE: 0.1715
MAE: 0.0140


# Decision Tree Hyperparameter Tuning kodi

In [15]:
from sklearn.model_selection import GridSearchCV

# ===============================
# Hyperparameterlar to‘plami
# ===============================
param_grid = {
    "max_depth": [5, 10, 15, 20, None],
    "min_samples_split": [2, 5, 10],
    "min_samples_leaf": [1, 2, 4],
    "max_features": [None, "sqrt", "log2"]
}

# ===============================
# GridSearchCV yaratish
# ===============================
grid_search = GridSearchCV(
    estimator=DecisionTreeRegressor(random_state=42),
    param_grid=param_grid,
    scoring="r2",
    cv=5,
    n_jobs=-1,
    verbose=1
)

# ===============================
# Tuningni bajarish
# ===============================
grid_search.fit(X_train, y_train)

# ===============================
# Eng yaxshi parametrlar va model natijasi
# ===============================
print("📌 Eng yaxshi parametrlar:", grid_search.best_params_)
print(f"📌 Eng yaxshi R2 (train): {grid_search.best_score_:.4f}")

# ===============================
# Eng yaxshi modelni test qilish
# ===============================
best_model = grid_search.best_estimator_
y_pred_tuned = best_model.predict(X_test)

r2_tuned = r2_score(y_test, y_pred_tuned)
mse_tuned = mean_squared_error(y_test, y_pred_tuned)
mae_tuned = mean_absolute_error(y_test, y_pred_tuned)

print("\n📌 Tuned Decision Tree Regressor — Natijalar:")
print(f"R2: {r2_tuned:.4f}")
print(f"MSE: {mse_tuned:.4f}")
print(f"MAE: {mae_tuned:.4f}")


Fitting 5 folds for each of 135 candidates, totalling 675 fits
📌 Eng yaxshi parametrlar: {'max_depth': 5, 'max_features': 'sqrt', 'min_samples_leaf': 1, 'min_samples_split': 5}
📌 Eng yaxshi R2 (train): 0.2401

📌 Tuned Decision Tree Regressor — Natijalar:
R2: -3.3097
MSE: 2.5356
MAE: 0.2562


In [16]:
from sklearn.model_selection import RandomizedSearchCV
from sklearn.tree import DecisionTreeRegressor
import numpy as np

# ===============================
# Hyperparameterlar oralig‘i
# ===============================
param_dist = {
    "max_depth": [None] + list(np.arange(5, 51, 5)),   # None yoki 5 dan 50 gacha 5 qadam bilan
    "min_samples_split": list(np.arange(2, 21, 2)),    # 2 dan 20 gacha
    "min_samples_leaf": list(np.arange(1, 11, 1)),     # 1 dan 10 gacha
    "max_features": [None, "sqrt", "log2"]
}

# ===============================
# RandomizedSearchCV yaratish
# ===============================
random_search = RandomizedSearchCV(
    estimator=DecisionTreeRegressor(random_state=42),
    param_distributions=param_dist,
    n_iter=50,           # sinov soni
    scoring="r2",        # baholash kriteriyasi
    cv=5,                # K-Fold
    verbose=1,
    random_state=42,
    n_jobs=-1
)

# ===============================
# Tuning jarayonini bajarish
# ===============================
random_search.fit(X_train, y_train)

# ===============================
# Eng yaxshi parametrlar va natijalar
# ===============================
print("📌 Eng yaxshi parametrlar:", random_search.best_params_)
print(f"📌 Eng yaxshi R2 (train): {random_search.best_score_:.4f}")

# ===============================
# Eng yaxshi modelni test qilish
# ===============================
best_model = random_search.best_estimator_
y_pred_tuned = best_model.predict(X_test)

from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
r2_tuned = r2_score(y_test, y_pred_tuned)
mse_tuned = mean_squared_error(y_test, y_pred_tuned)
mae_tuned = mean_absolute_error(y_test, y_pred_tuned)

print("\n📌 Tuned Decision Tree Regressor — Natijalar:")
print(f"R2: {r2_tuned:.4f}")
print(f"MSE: {mse_tuned:.4f}")
print(f"MAE: {mae_tuned:.4f}")


Fitting 5 folds for each of 50 candidates, totalling 250 fits
📌 Eng yaxshi parametrlar: {'min_samples_split': np.int64(4), 'min_samples_leaf': np.int64(2), 'max_features': 'log2', 'max_depth': np.int64(5)}
📌 Eng yaxshi R2 (train): 0.2266

📌 Tuned Decision Tree Regressor — Natijalar:
R2: -0.6391
MSE: 0.9644
MAE: 0.2285


# 📊 Tuned Decision Tree Regressor — Taqqoslash Jadvali

| Model                              | Eng yaxshi parametrlar                                                                        | Train R² | Test R²  | MSE     | MAE     |
|------------------------------------|------------------------------------------------------------------------------------------------|----------|----------|---------|---------|
| RandomizedSearchCV (n_iter=135)  | {'max_depth': 5, 'max_features': 'sqrt', 'min_samples_leaf': 1, 'min_samples_split': 5}       | 0.2401   | -3.3097  | 2.5356  | 0.2562  |
| RandomizedSearchCV (n_iter=50)   | {'min_samples_split': 4, 'min_samples_leaf': 2, 'max_features': 'log2', 'max_depth': 5}       | 0.2266   | -0.6391  | 0.9644  | 0.2285  |

---

## 🔍 Xulosa

1. **Train R²** ikkinchi tuningda biroz past, ammo **test natija** jihatidan yaxshiroq (Test R²: -0.6391 vs -3.3097).  
   Bu ikkinchi tuning modelning umumiylashish qobiliyatini yaxshilashga urinayotganini ko‘rsatadi.

2. **MSE va MAE** qiymatlari ikkinchi tuningda ancha past — bu modelning test ma’lumotlari bo‘yicha xatoliklari kamayganini bildiradi.

3. Har ikkala modelda Test R² manfiy bo‘lishi, model hali ham yaxshi umumiylashmaganini bildiradi — bu data structure va feature engineering masalasini qayta ko‘rib chiqishni talab qiladi.


In [17]:
import joblib
import os

# Modelni saqlash papkasi
model_folder = r"C:\Users\Rasulbek_Ruzmetov\Desktop\SML_R_Project\Models"

# Agar papka mavjud bo‘lmasa, yaratish
if not os.path.exists(model_folder):
    os.makedirs(model_folder)

# Fayl yo‘li
model_path = os.path.join(model_folder, "Decision_Tree_Model.pkl")

# Modelni saqlash
joblib.dump(model, model_path)
print(f"✅ Model saqlandi: {model_path}")


✅ Model saqlandi: C:\Users\Rasulbek_Ruzmetov\Desktop\SML_R_Project\Models\Decision_Tree_Model.pkl
